src/H5FDsubfiling/H5subfiling_common.c

```diff
@@ -2088,24 +2082,40 @@ record_fid_to_subfile(uint64_t file_id, int64_t subfile_context_id, int *next_in
  * Purpose:     Remove the map entry associated with the file->inode.
  *              This is done at file close.
  *
- * Return:      None
- * Errors:      Cannot fail.
+ * Return:      Non-negative on success/Negative on failure
  *
  *-------------------------------------------------------------------------
  */
-static void
+static herr_t
 clear_fid_map_entry(uint64_t file_id, int64_t sf_context_id)
 {
-    if (sf_open_file_map) {
-        for (int i = 0; i < sf_file_map_size; i++) {
-            if ((sf_open_file_map[i].file_id == file_id) &&
-                (sf_open_file_map[i].sf_context_id == sf_context_id)) {
-                sf_open_file_map[i].file_id       = UINT64_MAX;
-                sf_open_file_map[i].sf_context_id = -1;
-                return;
-            }
+    herr_t ret_value = SUCCEED;
+
+    if (!sf_open_file_map)
+        H5_SUBFILING_GOTO_DONE(SUCCEED);
+
+    for (int i = 0; i < sf_file_map_size; i++) {
+        subfiling_context_t *sf_context = NULL;
+
+        if ((sf_open_file_map[i].file_id != file_id) || (sf_open_file_map[i].sf_context_id != sf_context_id))
+            continue;
+
+        /* Only clear map entry if this is the last file
+         * holding a reference to the context
+         */
+        if (NULL == (sf_context = H5_get_subfiling_object(sf_context_id)))
+            H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTGET, FAIL, "couldn't get subfiling context object");
+
+        if (sf_context->file_ref == 0 || sf_context->file_ref == 1) {
+            sf_open_file_map[i].file_id       = UINT64_MAX;
+            sf_open_file_map[i].sf_context_id = -1;
         }
```

Comment:
So, this could potentially happen twice, when file_ref ==1 then later file_ref == 0?  Just curious...

Reply:
In the current code with this PR, this function's only called in two places: either when opening a file fails where the object will have a ref. of 0 or during file close when the object will have a ref. of 1 and be freed shortly after calling this function. Code changes in the future could change things, but in general we shouldn't ever arrive here twice for the same context object. Usually you'd start a reference count on an object at 1, but dealing with proper reference counting in failure states can then be difficult.